cpp/src/arrow/util/ree_util.h

```diff
@@ -209,18 +209,19 @@ class RunEndEncodedArraySpan {
   /// \brief Create an iterator from a logical position and its
   /// pre-computed physical offset into the run ends array
   ///
-  /// \param logical_pos is an index in the [0, length()) range
+  /// \param logical_pos is an index in the [0, length()] range
   /// \param physical_offset the pre-calculated PhysicalIndex(logical_pos)
   Iterator iterator(int64_t logical_pos, int64_t physical_offset) const {
     return Iterator{PrivateTag{}, *this, logical_pos, physical_offset};
   }
 
   /// \brief Create an iterator from a logical position
   ///
-  /// \param logical_pos is an index in the [0, length()) range
+  /// \param logical_pos is an index in the [0, length()] range
   Iterator iterator(int64_t logical_pos) const {
-    assert(logical_pos < length());
-    return iterator(logical_pos, PhysicalIndex(logical_pos));
+    return iterator(logical_pos, logical_pos < length()
+                                     ? PhysicalIndex(logical_pos)
+                                     : RunEndsArray(array_span).length);
```

Comment:
was this caught by a test?

Reply:
I tried to use this class for something new and I realized this assert was too restrictive: it's not necessary for the correct working of the class.

Reply:
`PhysicalIndex(logical_pos)` triggers an assert if a bad index is passed, so that's why I have the conditional now.